### DISCUSSION TITLE
Auto-import should have context-aware sorting rules

### DISCUSSION THREAD

**Discussion by Prunoideae** [microsoft/pylance-release#2579] [Enhancement]
Auto import is a great feature, but sometimes it become rather annoying if a short keyword argument have a same name in any package. For example:

![b](https://user-images.githubusercontent.com/24620047/94336409-b53e3c00-0015-11eb-8bd4-ae4ec1336878.gif)

Exact match for keywords should have a higher priority, but I think there should be some context priority too, like in this time, the 'sep=' item should appear before the auto-import of 'os.sep' occurs.

**Comment by Prunoideae** [comment]
Another example is:

![b](https://user-images.githubusercontent.com/24620047/94357299-55559d00-00ca-11eb-94f4-81af51db11da.gif)

**Comment by jakebailey** [comment]
This doesn't fix the problem, but one thing you may want to do is ensure your completion tooltip is expanded; you appear to have the sidebar hidden which means the info we're trying to show about what will be inserted isn't being shown.

**Comment by jakebailey** [comment]
The next release will also put an indicator into the completion list, even with the docs hidden:

![image](https://user-images.githubusercontent.com/5341706/94482742-9e6f2500-018e-11eb-8ed1-056dee6cba52.png)

**Comment by heejaechang** [comment]
the issue itself can be fixed, but making our features to understand context is a lot bigger work since we don't track context at all currently.

but that being said for the "final" example above, even if we make us to context-aware, I don't think we can fix that particular one. final in typing is defined as a function "def final ..." and the fact that it is supposed to be used as a decorator is not encoded in any way except in doc comment and that is not something we can support unless we hard code it that we don't want to do since it won't work for any other decorator.

**Reply by Bobronium** [reply]
How's the "decorator vs function" discussion relevant here though?

The issue with this example is that Pylance suggests to import anything at all when there's a perfect match for completion in local scope (`final` -> `finals`).

Where is the write place to write a proposal on how it should work?

**Comment by heejaechang** [comment]
I will leave it open in case we want to make all our features context aware for example, don't show invalid keywords in completion (showing "import" keyword in expression context)

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by Bobronium** [comment]
This is really annoying:
<img width="600" alt="image" src="https://user-images.githubusercontent.com/36469655/169266601-355792fc-b119-4670-a249-57162981313c.png">
and inconsistent as well (notice first screenshot doesn't have "foo" in suggestions, and this one does):
<img width="600" alt="image" src="https://user-images.githubusercontent.com/36469655/169266825-fcfc9650-f05f-493f-bae3-504cb836f478.png">

How it should behave (screenshot from PyCharm):
<img width="573" alt="image" src="https://user-images.githubusercontent.com/36469655/169267263-1d76ed70-7e2f-4e14-8113-c1bd1f57a962.png">

Imports suggested by pylance also don't make much sense: they should be case-sensitive, following python syntax rules.

When/will this issue be prioritised?
